### PR TITLE
Fix import queue thread pool shutdown

### LINKED ISSUE #4913: check-block sometimes Segmentation fault: 11
```
$ ./target/debug/substrate check-block --dev 2 --pruning=archive
2020-02-13 22:59:06 Importing 1 blocks
2020-02-13 22:59:08 Imported 1 blocks. Best: #6
Completed in 1256 ms.
Segmentation fault: 11
```

But seems working regardless. I guess some exit handler not configured correctly?
Also saw a few instance where it didn't seg fault, so some timing / ordering issue.

And it will complain pruning mode is not achieve if I didn't specify `--pruning=archive`, which seems redundant.

### PR DESCRIPTION
**Pull request by arkpar**
Due to the way rs-futures `ThreadPool` is implemented, Importing threads may outlive the queue instance and even the main thread. This leads to a race on closing rocksdb database vs cleaning the C++ runtime in the main thread, causing segfaults and preventing the database from being closed properly.

This PR adds explicit synchronization between importing threads and the queue shutdown.

Fixes #4913

### CODE REVIEW

File: primitives/consensus/common/src/import_queue/basic_queue.rs [hunk 1]
```diff
@@ -40,9 +41,29 @@ pub struct BasicQueue<B: BlockT, Transaction> {
...
 	_phantom: PhantomData<Transaction>,
 }
 
+impl<B: BlockT, Transaction> Drop for BasicQueue<B, Transaction> {
+	fn drop(&mut self) {
+		drop(self.pool.take());
+		// Flush the queue and close the receiver to terminate the future.
+		let _ = self.sender.unbounded_send(ToWorkerMsg::Shutdown);
+		let (_, closed) = buffered_link::buffered_link();
+		drop(std::mem::replace(&mut self.result_port, closed));
```
Comment (bkchr) [inline primitives/consensus/common/src/import_queue/basic_queue.rs]: Why do we need to do this?

File: primitives/consensus/common/src/import_queue/basic_queue.rs [hunk 2]
```diff
@@ -40,9 +41,29 @@ pub struct BasicQueue<B: BlockT, Transaction> {
...
 	pool: Option<futures::executor::ThreadPool>,
+	pool_guard: Arc<(Mutex<usize>, Condvar)>,
 	_phantom: PhantomData<Transaction>,
 }
 
+impl<B: BlockT, Transaction> Drop for BasicQueue<B, Transaction> {
+	fn drop(&mut self) {
+		drop(self.pool.take());
+		// Flush the queue and close the receiver to terminate the future.
+		let _ = self.sender.unbounded_send(ToWorkerMsg::Shutdown);
```
Comment (bkchr) [inline primitives/consensus/common/src/import_queue/basic_queue.rs]: I think we should just close it: https://docs.rs/futures/0.3.4/futures/channel/mpsc/struct.UnboundedSender.html#method.close_channel

File: primitives/consensus/common/src/import_queue/basic_queue.rs [hunk 3]
```diff
@@ -63,9 +84,22 @@ impl<B: BlockT, Transaction: Send + 'static> BasicQueue<B, Transaction> {
...
 		);
 
+		let guard = Arc::new((Mutex::new(0usize), Condvar::new()));
+		let guard_start = guard.clone();
+		let guard_end = guard.clone();
+
 		let mut pool = futures::executor::ThreadPool::builder()
 			.name_prefix("import-queue-worker-")
-			.pool_size(1)
+			.pool_size(2)
```
Comment (bkchr) [inline primitives/consensus/common/src/import_queue/basic_queue.rs]: Why do we increase the size?

Comment (bkchr) [reply primitives/consensus/common/src/import_queue/basic_queue.rs]: Isn't there just one worker anyway?

Comment (arkpar) [reply primitives/consensus/common/src/import_queue/basic_queue.rs]: Was testing if thread tracking code is working properly. Eventually there should be more workers, but this is out of scope of this PR.

**Comment by arkpar** [comment]
Also fixed an issue with slog global guard being disposed too early.

File: primitives/consensus/common/src/import_queue/basic_queue.rs (hunk 1 above)
Comment (arkpar) [reply primitives/consensus/common/src/import_queue/basic_queue.rs]: Replaces with `close`

File: primitives/consensus/common/src/import_queue/basic_queue.rs [hunk 4]
```diff
@@ -144,6 +178,7 @@ enum ToWorkerMsg<B: BlockT> {
 	ImportBlocks(BlockOrigin, Vec<IncomingBlock<B>>),
 	ImportJustification(Origin, B::Hash, NumberFor<B>, Justification),
 	ImportFinalityProof(Origin, B::Hash, NumberFor<B>, Vec<u8>),
+	Shutdown,
```
Comment (bkchr) [inline primitives/consensus/common/src/import_queue/basic_queue.rs]:
```suggestion
```

File: primitives/consensus/common/src/import_queue/basic_queue.rs [hunk 5]
```diff
@@ -239,6 +274,7 @@ impl<B: BlockT, Transaction: Send> BlockImportWorker<B, Transaction> {
 					ToWorkerMsg::ImportJustification(who, hash, number, justification) => {
 						worker.import_justification(who, hash, number, justification);
 					}
+					ToWorkerMsg::Shutdown => return Poll::Ready(()),
```
Comment (bkchr) [inline primitives/consensus/common/src/import_queue/basic_queue.rs]:
```suggestion
```

File: primitives/consensus/common/src/import_queue/basic_queue.rs (hunk 2 above)
Comment (bkchr) [reply primitives/consensus/common/src/import_queue/basic_queue.rs]:
```suggestion
		self.sender.close_channel();
```

File: primitives/consensus/common/src/import_queue/basic_queue.rs [hunk 6]
```diff
@@ -40,9 +41,28 @@ pub struct BasicQueue<B: BlockT, Transaction> {
 	manual_poll: Option<Pin<Box<dyn Future<Output = ()> + Send>>>,
 	/// A thread pool where the background worker is being run.
 	pool: Option<futures::executor::ThreadPool>,
+	pool_guard: Arc<(Mutex<usize>, Condvar)>,
```
Comment (ordian) [inline primitives/consensus/common/src/import_queue/basic_queue.rs]: Any reason not to use a https://docs.rs/crossbeam/0.7.3/crossbeam/sync/struct.WaitGroup.html?

Comment (arkpar) [reply primitives/consensus/common/src/import_queue/basic_queue.rs]: 1. There's no control over the lifetimes of objects passed to `ThreadPoolBuilder`. It uses some reference counting internally and relying on implementation details seems wrong.
2. Don't want to introduce another dependency over a trivial thing.

**Pull request merged**